검색 도메인 리팩토링, 최근 검색 기록 구현

vectorch9 opened the pull request:
<!--
✅ Resolve: #이슈번호 형태로 입력해 주세요.
ex) Resolve: #123
-->
## 📌Linked Issues
- #162 

<!--
✅ 변경 사항을 자세히 알려주세요. (why -> what -> how)
-->
## ✏Change Details
### Search 도메인 리팩토링
- 컨벤션 적용
- searchDTO의 의미 없는 값(count) 삭제 -> 쿼리 조회 시 orderBy로 정렬되기 때문에 사용하지 않음
- SearchRepositoryQuerydslImpl의 복잡한 쿼리 구조 개선
  - 내부 로직을 분리하여 다시 활용할 수 있도록함 -> 이후 최근 검색 조회 시에도 사용 
  - 검색 조회 결과에 검색 키워드와 완전히 일치하는 유저,해시태그가 강제로 추가되는 경우 정상작동 하지 않던 일부 로직 정상화
- EntityTypeInvalidException 추가

### 최근 검색 API 추가
- 최근 검색 기록 조회 API 추가
- 최근 검색 기록 삭제, 전체 삭제 API 추가
- 기존의 조회수 증가 API에 최근 검색기록을 업데이트 하는 기능 추가
<!--
✅ 추가로 전달할 내용이 있다면 적어주세요.
-->
## 💬Comment
- 코드 정렬 때문에 `followRepository`도 변경사항이 많아보이는데, `getFollowingMemberFollowMap` 부분만 추가되었습니다.
 -> 원래 검색 도메인에서 처리하던 내용을 분리했습니다.
- 아직 미완성인 기능이 있습니다. 관련 Issue #163 

<!--
✅ 참고한 사이트가 있다면 공유해주세요.
-->
## 📑References
- https://cjh5414.github.io/jpa-use-composite-key-to-foreign-key/
  - `RecentSearch` 구현시 위 사이트를 참고하여 두개의 FK를 묶어 복합키(PK)로 사용했습니다.
  - `Follow`와 같은 엔티티도 위의 형식으로 바꾸면 좋을거 같은데 어떻게 생각하시나요?
> `Follow`와 같은 경우 이미 두 `member`를 통해 유일성이 보장되기 때문에 추가적인 `id`가 필요없을 것 같습니다 !

## ✅Check List
- [x] 추가한 기능에 대한 테스트는 모두 완료하셨나요?
- [x] 코드 정렬(Ctrl + Alt + L), 불필요한 코드나 오타는 없는지 확인하셨나요?

vectorch9 committed:
Refactor: Search 도메인 리팩토링

- 컨벤션 적용
- searchDTO의 의미 없는 값(count) 삭제 -> 쿼리 조회 시 orderBy로 정렬되기 때문에 사용하지 않음
- SearchRepositoryQuerydslImpl의 복잡한 쿼리 구조 개선
  - 내부 로직을 분리하여 다시 활용할 수 있도록함
  - 검색 조회 결과에 검색 키워드와 완전히 일치하는 유저,해시태그가 강제로 추가되는 경우 정상작동 하지 않던 일부 로직 정상화
- EntityTypeInvalidException 추가

vectorch9 committed:
Feature: 최근 검색 API 추가

- 최근 검색 기록 조회 API 추가
- 최근 검색 기록 삭제, 전체 삭제 API 추가
- 기존의 조회수 증가 API에 최근 검색기록을 업데이트 하는 기능 추가

seonpilKim commented on src/main/java/cloneproject/Instagram/domain/follow/repository/querydsl/FollowRepositoryQuerydslImpl.java:
```diff
@@ -1,91 +1,114 @@
+
+	@Override
+	public List<FollowerDTO> getFollowings(Long loginedMemberId, Long memberId) {
+
```
메소드 첫 줄 공백 제거해 주시면 될 것 같아요

vectorch9 replied:
본문에서 언급한 것 처럼 추가된 메소드만 검토하고 나머진 자동정렬이라 빼먹었네요..

seonpilKim commented on src/main/java/cloneproject/Instagram/domain/search/controller/SearchController.java:
```diff
@@ -27,30 +30,66 @@
 @RestController
 @RequiredArgsConstructor
 public class SearchController {
```
컨트롤러단에 `@RequestMapping("/topsearch")`으로 빼면 좋을 것 같아요.
ResultCode도 static import해서 사용하면 좀 더 편리할 것 같아요.

vectorch9 replied:
수정하겠습니다.

seonpilKim commented on src/main/java/cloneproject/Instagram/domain/search/entity/RecentSearch.java:
```diff
@@ -0,0 +1,61 @@
+@Entity
+@Table(name = "recent_searches")
+@IdClass(RecentSearch.RecentSearchId.class)
+public class RecentSearch {
```
저는 MySQL에서는 PK를 복합 키로 잘 사용하지 않는 것으로 알고 있습니다.
> Reference: [우아한형제 기술이사 김영한](https://www.inflearn.com/questions/38108), 스프링 부트와 AWS로 혼자 구현하는 웹 서비스(이동욱) p.91

#### PK를 복합 키로 잘 사용하지 않는 이유
1. FK를 맺을 때, 다른 테이블에서 복합 키를 모두 갖고 있거나, 중간 테이블을 두어야 하는 공간 낭비
2. 인덱스에 좋은 영향을 끼치지 못함 -> 쿼리 성능 저하
3. 유니크한 조건이 변경될 경우, PK 전체를 수정해야 하는 상황 발생
4. JPA의 영속성 컨텍스트에서 해당 엔티티의 캐시 키 사이즈 증가

#### PK로 Auto Increment를 사용하는 이유
1. PK를 UUID나 복합 키로 사용하는 경우, insert마다 재정렬이 발생하지만, Auto Increment는 재정렬이 필요 없으므로 성능상 이점이 있음
2. Storage 측면에서 이점이 있음

PK를 복합 키로 사용하는 것이 사람 입장에서 좀 더 자연스럽고, 쿼리 측면에서 조금 단순해질 수 있지만, 꼭 복합 키로 묶지 않고도 인덱스나 제약 조건을 사용해도 문제 없다는 걸로 알고 있습니다.

일단 제 생각을 말씀드려 봤는데.. 다훈님은 혹시 PK를 복합 키로 사용하신 다른 이유가 있으실까요?

vectorch9 replied:
우선 복합키를 사용하려 했던 두가지 이유가 있습니다.
1. `RecentSearch`는 select 쿼리 실행시 조건절에  항상 member가 들어갑니다. (해당 멤버가 조회한 기록을 불러오기 때문입니다.)  mysql InnoDB의 Stroage는 PK를 기준으로 clustering이 되어 저장되기 때문에 member를 PK에 포함하면 select 쿼리 실행 시 이점이 있다고 생각합니다. (따라서 복합키의 경우 컬럼순서가 중요하다고 합니다.) 대신 insert, delete 시에는 재정렬이 필요하기 때문에 약간의 성능저하가 있습니다. 거꾸로 Auto increment PK를 사용시에는 insert, delete시엔 재정렬이 필요하지 않지만 select 쿼리 조회 시 member를 기준으로 clsutering 된 것이아닌 ID를 기준으로 clustering이 되었기 때문에 member의 최근 검색 기록을 모두 불러오는 쿼리 실행시 성능 저하가 발생합니다. 

> select쿼리는 매 검색시 발생하기 때문에 좀 더 최적화가 필요한 부분이라고 생각합니다. 검색창에 커서를 올리는 순간 최근검색 API는 무조건 호출되는 방식이라 `select 쿼리 호출 횟수` >= `insert, delete 쿼리 호출 횟수`가 되어 select 쿼리의 실행속도가 좀더 중요하다고 생각합니다.

2. 논쟁이 많은 부분인데, 두 테이블 사이의 many to many 관계를 표현할 때는 별도의 PK가 필요없다고 생각했습니다. 이 부분은 괄호스타일의 K&R과 BSD처럼 각자의 스타일과 컨벤션이 반영되는 부분이지만, 1번의 이유를 함께 고려해 별도의 PK (autoincrement id)가 없는 형태로 구현을 해보았습니다.

seonpilKim replied:
다훈님 말씀처럼 조회 측면에서 복합 키를 PK로 사용하면 이점이 있겠네요.

사실 개인적으로 복합 키를 PK로 사용하면 생기는 개발상의 단점들 때문에, 가능하면 PK는 auto incre로 사용하고 필요하면 제약조건이나 인덱스를 선호하는 편이라, 고민이 되네요..

인덱스를 사용하는 것과, PK에 복합 키를 사용하는 것의 전반적인 성능 차이가 얼마나 나는지 테스트나 다른 자료들도 좀 더 비교해보고 싶은데 시간이 도와주질 않네요😵

일단 이 방식으로 사용해 보고, 시험 끝난 다음에 한 번 더 고민해 보면 좋을 것 같아요.

seonpilKim commented on src/main/java/cloneproject/Instagram/domain/search/repository/querydsl/RecentSearchRepositoryQuerydsl.java:
```diff
@@ -0,0 +1,18 @@
+import cloneproject.Instagram.domain.search.entity.RecentSearch;
+import cloneproject.Instagram.domain.search.entity.Search;
+
+public interface RecentSearchRepositoryQuerydsl {
```
클래스 첫 줄과 메소드 사이에 새 줄 추가해 주시면 될 것 같아요

seonpilKim commented on src/main/java/cloneproject/Instagram/domain/search/repository/querydsl/SearchRepositoryQuerydslImpl.java:
```diff
@@ -1,258 +1,147 @@
+
+	@Override
+	public List<Search> findAllByTextLike(String text) {
+		String keyword = text + "%";
```
final 키워드 추가해 주시면 좋을 것 같아요 (다른 파일들도 확인해 주세요)

seonpilKim commented on src/main/java/cloneproject/Instagram/domain/search/service/SearchService.java:
```diff
@@ -1,71 +1,197 @@
+	private static final int PAGE_SIZE = 5;
+	private static final int PAGE_OFFSET = 2;
+
+	@Transactional(readOnly = true)
```
`@Transactional(readOnly = true)`은 클래스 단위에 추가하고, 데이터 변경이 있는 메소드에만 `@Transactional`을 명시해 주는 방식으로 코드를 작성하면 좀 더 편리하게 하실 수 있을 것 같아요.

seonpilKim commented on src/main/java/cloneproject/Instagram/domain/search/service/SearchService.java:
```diff
@@ -1,71 +1,197 @@
+	private final RecentSearchRepository recentSearchRepository;
+	private final MemberStoryRedisRepository memberStoryRedisRepository;
+	private final FollowRepository followRepository;
+	private static final int FIRST_PAGE_SIZE = 15;
```
구분을 위해 static변수 위에 새 줄 추가해 주면 좋을 것 같아요.

seonpilKim commented on src/main/java/cloneproject/Instagram/domain/search/service/SearchService.java:
```diff
@@ -1,71 +1,197 @@
+				.member(loginMember)
+				.search(search)
+				.build());
+		recentSearch.updateLastSearchedAt();
```
저희 DB 컨벤션이 create_date, update_date 이라 `~At` 보다는 `~Date`로 네이밍하는 게 좋을 것 같아요.

vectorch9 replied:
위의 컨벤션과 코드스타일 부분은 수정하겠습니다. 시험기간이라 급하게 작업하다보니 검토가 덜된 부분이 많네요.. 😓

seonpilKim commented on src/main/java/cloneproject/Instagram/global/error/exception/EntityTypeInvalidException.java:
```diff
@@ -0,0 +1,9 @@
+
+import cloneproject.Instagram.global.error.ErrorCode;
+
+public class EntityTypeInvalidException extends BusinessException{
```
이 파일에는 컨벤션 적용이 전체적으로 안 되어 있네요

vectorch9 committed:
Style: Search 도메인 컨벤션 적용

- 빈줄 컨벤션 적용
- lastSearchedAt을 lastSearchedDate으로 수정
- SearchController의 /topsearch부분을 묶어서 controller에 매핑
- @Transactional(readOnly=true)를 Service 클래스에 적용

seonpilKim reviewed (approved):
👍
